### ISSUE TITLE
Parameterize small molecules

### ISSUE THREAD

**Issue by fwaibl** [hits-mbm-dev/grappa#18]
Hi. Thanks for making GrAPPa open-source! I wanted to try it for some small-molecule systems in a GROMACS workflow. Is that possible, and what would be the best strategy to do so?

Here's what I tried so far:

# Via Gromacs topology
For this, I used a GaFF 2 topology converted from Amber parm7 to Gromacs .top via parmed. The problem seems to be that the resulting topology has all terms defined directly in the [bonds], [angles] etc. sections, while protein topologies usually have [bondtypes], [angletypes] etc. that contain the parameters per residue. However, I don't know a straight-forward way of getting my small-molecule topology into this format.

This is my error message:
```
Topology does not contain section bondtypes. Is the forcefield in the correct directory?
Topology does not contain section angletypes. Is the forcefield in the correct directory?
Topology does not contain section dihedraltypes. Is the forcefield in the correct directory?
ffdir is None. No residuetypes will be parsed.
topology does not contain section bondtypes
topology does not contain section angletypes
Parameterizing molecule without focus
Traceback (most recent call last):
  File "/localhome/fwaibl/.conda/envs/grappa/bin/grappa_gmx", line 8, in <module>
    sys.exit(main())
  File "/localhome/fwaibl/programs/grappa/src/grappa/wrappers/gromacs_wrapper.py", line 81, in main
    return main_(args.top_path, top_outpath=args.top_outpath, modeltag=args.modeltag, charge_model=args.charge_model, device=args.device, plot_parameters=args.plot_parameters)
  File "/localhome/fwaibl/programs/grappa/src/grappa/wrappers/gromacs_wrapper.py", line 68, in main_
    grappa.parametrize(top_path, top_outpath, charge_model=charge_model, plot_parameters=plot_parameters)
  File "/localhome/fwaibl/programs/grappa/src/grappa/wrappers/gromacs_wrapper.py", line 61, in parametrize
    write_top(topology.to_dict(), top_outpath)
  File "/localhome/fwaibl/.conda/envs/grappa/lib/python3.10/site-packages/kimmdy/topology/topology.py", line 964, in to_dict
    self.update_parameters()
  File "/localhome/fwaibl/.conda/envs/grappa/lib/python3.10/site-packages/kimmdy/topology/topology.py", line 866, in update_parameters
    self.parametrizer.parameterize_topology(self, focus_nrs)
  File "/localhome/fwaibl/programs/grappa/src/grappa/utils/kimmdy_utils.py", line 274, in parameterize_topology
    mol = build_molecule(current_topology, build_nrs, charge_model=self.charge_model)
  File "/localhome/fwaibl/programs/grappa/src/grappa/utils/kimmdy_utils.py", line 82, in build_molecule
    atom_info["atomic_number"].append(int(at_map[atom.type].at_num))
ValueError: invalid literal for int() with base 10: 'f'
```

And this was my input file:
```
grappa_gmx -f g16-amber.top -o g16-grappa.top -t grappa-1.2
```

# Via OpenFF tools
For this, I made a new conda environment and ran `bash installation_openmm.sh` to install grappa. I then load a small molecule from a mol file into an OpenFF Molecule, and parameterize with OpenFF to get an OpenMM system. When I run grappa, it runs for a while, and then produces the following error:
```
Traceback (most recent call last):
  File "/home/fwaibl/Documents/test-grappa/openmm/run-grappa.py", line 21, in <module>
    system = grappa_ff.parametrize_system(system, topology, charge_model='am1BCC')
  File "/localhome/fwaibl/programs/grappa/src/grappa/wrappers/openmm_wrapper.py", line 42, in parametrize_system
    sub_topology = get_subtopology(topology, exclude_residues=exclude_residues)
  File "/localhome/fwaibl/programs/grappa/src/grappa/utils/openmm_utils.py", line 32, in get_subtopology
    for atom in topology.atoms():
TypeError: 'generator' object is not callable
```

This is my full input:
```
from openmm.app import Topology
from openff.toolkit import Molecule, ForceField
from grappa import OpenmmGrappa
from openff.interchange import Interchange

ligand = Molecule.from_file('g16-rdkit.mol')
topology = ligand.to_topology()

classical_ff = ForceField('openff-1.2.0.offxml')                                    
interchange: Interchange = Interchange.from_smirnoff(
    force_field=classical_ff, topology=topology
)
system = interchange.to_openmm_system()
# system = classical_ff.createSystem(topology)                                      

# load the pretrained ML model from a tag. Currently, possible tags are 'grappa-1.1', grappa-1.2' and 'latest'
grappa_ff = OpenmmGrappa.from_tag('grappa-1.2')

# parametrize the system using grappa.
# The charge_model tag tells grappa how the charges were obtained, in this case from the classical forcefield amberff99sbildn. possible tags are 'amber99' and 'am1BCC'. 

# this is where the error occurs:
system = grappa_ff.parametrize_system(system, topology, charge_model='am1BCC')
```

**Comment by ehhartmann** [comment]
Hi Franz, 
thanks for trying out grappa! What you want to do is within the scope of grappa and your approach looks good so far. We haven't tried topologies from parmed, yet and I guess there are some problems with the Atomtype definition. Not having bondtypes etc is fine because grappa can deal with the explicit definition of bond parameters in the bonds section.
 We expect atomtypes to be defined like this:
'''
[ atomtypes ]
; name      at.num  mass     charge ptype  sigma      epsilon
Br          35      79.90    0.0000  A   0.00000e+00  0.00000e+00
C            6      12.01    0.0000  A   3.39967e-01  3.59824e-01
CA           6      12.01    0.0000  A   3.39967e-01  3.59824e-01
'''
Could it be that the second entry in your atomtypes definition is not the atomic number? 
Best
Eric

**Comment by fwaibl** [comment]
Dear Eric,

thanks for your response. Indeed, my topology has the `bond_type` in the second column. If I replace this column by the atomic number, `grappa_gmx` runs successfully, but still prints many warnings. I append the warnings below. As far as I can see, all parameters are generated (bonds, angles, proper and improper dihedrals). Do you think that the generated topology is ok?

```
Topology does not contain section bondtypes. Is the forcefield in the correct directory?                         
Topology does not contain section angletypes. Is the forcefield in the correct directory?                        
Topology does not contain section dihedraltypes. Is the forcefield in the correct directory?                     
ffdir is None. No residuetypes will be parsed.                                                                   
topology does not contain section bondtypes                                                                      
topology does not contain section angletypes                                                                     
Parameterizing molecule without focus                                                                            
WARNING:root:Ignored parameters with invalid ids: ('20', '15', '16', '17') for proper dihedrals                  
WARNING:root:Ignored parameters with invalid ids: ('20', '15', '16', '19') for proper dihedrals                  
WARNING:root:Ignored parameters with invalid ids: ('20', '22', '46', '17') for proper dihedrals
... many more equivalent lines
WARNING:kimmdy.topology.utils:topology does not contain section bondtypes                                        
WARNING:kimmdy.topology.utils:topology does not contain section angletypes    
```

**Comment by ehhartmann** [comment]
Hey,
I think the topology `Ignored parameters with invalid ids` error messages are a problem.Would you expect to have proper dihedrals for a set of the mentioned atom ids?

**Comment by fwaibl** [comment]
Thanks again for your help. I cannot check the dihedrals in detail now because I am at a conference, but will do so next week. From my preliminary checks, it seems that the rejected dihedrals should not exist in the topology, and actually GrAPPa has all dihedral terms that are in the input topology, and even a few more.

**Comment by fwaibl** [comment]
Hello,
I checked the dihedrals again and found the following: The warning seems to occur for dihedral angles that are specified in the inverse order (i.e., the first index is larger than the last one). The respective parameters are not modified, but stay the same as in the original topology.
As a further test, I added some code to `src/grappa/utils/kimmdy_utils.py`. To show my change: this is the output of `git diff` (the lines with a + are the one I added):
```
diff --git a/src/grappa/utils/kimmdy_utils.py b/src/grappa/utils/kimmdy_utils.py
index 50e5ec5..fc50df2 100644
--- a/src/grappa/utils/kimmdy_utils.py
+++ b/src/grappa/utils/kimmdy_utils.py
@@ -191,6 +191,9 @@ if importlib.util.find_spec('kimmdy') is not None:
         for i, idx in enumerate(parameters.propers):
             if all(atom_nr in apply_nrs for atom_nr in idx):
                 tup = tuple(idx)
+                tup_rev = tuple(reversed(idx))
+                if tup_rev in top.proper_dihedrals and tup not in top.proper_dihedrals:
+                    tup = tup_rev
                 if not top.proper_dihedrals.get(tup):
                     # raise KeyError(f"bad index {tup} in {list(top.proper_dihedrals.keys())}")
                     logging.warning(
@@ -284,4 +287,4 @@ if importlib.util.find_spec('kimmdy') is not None:
 
             # apply parameters
             apply_parameters(current_topology, parameters, build_nrs)
-            return current_topology
\ No newline at end of file
+            return current_topology
```
With this change, I do not get the warnings anymore, and the dihedral angles which previously gave me a warning are now changed as well.

What is your opinion? Do you think that my modified version of GrAPPa will give me correct results? Should I change anything else?

Best,
Franz

**Comment by LeifSeute** [comment]
Hi Franz,
thanks for pointing that out! It is indeed a bug of the gmx interface since grappa may change the index order of an interaction internally (to enforce a more standardized way of defining idx tuples) and we have to check whether this has happened.
What you suggested is the correct solution and I incorporated it along with analogous versions for bonds and angles in a bug fix #19.

The problem with the Openff/Openmm pipeline was that you passed an openff.Topology, but Grappa expects an openmm.app.Topology. I included a more informative error message for this case.

Thanks for using Grappa, these feedbacks help us a lot in developing the package!

Best,
Leif

**Comment by fwaibl** [comment]
Thanks a lot for your help and for the quick bugfix, I will update my installation.

> The problem with the Openff/Openmm pipeline was that you passed an openff.Topology, but Grappa expects an openmm.app.Topology. I included a more informative error message for this case.

Ok, that explains it!

Best,
Franz
